File: apps/api/src/test/kotlin/dev/vibemotion/api/export/ExportParityTest.kt
package dev.vibemotion.api.export

import dev.vibemotion.api.catalog.keyframesName
import dev.vibemotion.api.domain.Trigger
import io.kotest.assertions.withClue
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.ints.shouldBeGreaterThan
import io.kotest.matchers.shouldBe
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
private data class ParityCase(
    val variant: String,
    val params: Map<String, String>,
    val longhands: Map<String, String>,
    val customProperties: Map<String, String>,
)

@Serializable
private data class ParityEntry(
    val version: String,
    val animationId: String,
    val keyframesCss: String,
    val baseStyles: Map<String, String>,
    val cases: List<ParityCase>,
)

@Serializable
private data class ParityFixture(
    val generator: String,
    val entries: List<ParityEntry>,
)

private val parityJson = Json { ignoreUnknownKeys = true }

/** CSS initial values for the longhands the shorthand resets, in shorthand order after the name. */
private val INITIAL_LONGHANDS =
    listOf(
        "animation-duration" to "0s",
        "animation-timing-function" to "ease",
        "animation-delay" to "0s",
        "animation-iteration-count" to "1",
        "animation-direction" to "normal",
        "animation-fill-mode" to "none",
    )

/**
 * The editor preview renders CSS from TypeScript (`apps/web/lib/runtime-css`); the export renders
 * it from Kotlin. A designer who exports must get what the designer saw, so the two generators are
 * held to one shared artefact: `export-parity.json`, produced by the TypeScript side's own vitest
 * run for **every** published `(catalogVersion, animationId)` pair, in three param cases each.
 *
 * This reads the committed file off the test classpath (`exportParityResources` in
 * `build.gradle.kts`) and drives the real [CssEmitter] against it — the emitted bytes, parsed back,
 * not an internal shortcut.
 *
 * Triggers are not in the fixture: TypeScript has no trigger CSS. The real-browser specs in
 * `packages/bridge/e2e` and the full-stack spec in `apps/e2e/web/stack` cover those.
 */
class ExportParityTest :
    FunSpec({

        val fixtureJson =
            checkNotNull(
                ExportParityTest::class.java.classLoader.getResourceAsStream("export/export-parity.json"),
            ) {
                "Missing export/export-parity.json on the test classpath. Generate it with " +
                    "`VM_UPDATE_PARITY=1 pnpm --filter web test export-parity` and check that " +
                    "exportParityResources is wired into the test source set."
            }.bufferedReader()
                .use { it.readText() }
        val fixture = parityJson.decodeFromString(ParityFixture.serializer(), fixtureJson)

        val emitter = CssEmitter(CATALOG)

        test("the fixture covers exactly the published (version, animationId) pairs") {
            val published =
                CATALOG.versions().flatMap { version ->
                    CATALOG
                        .catalog(version)
                        ?.entries
                        .orEmpty()
                        .map { "$version/${it.id}" }
                }

            fixture.entries.map { "${it.version}/${it.animationId}" } shouldContainExactly published
            published.size shouldBeGreaterThan 0
        }

        test("every keyframes block is byte-identical to the preview's") {
            fixture.entries.forEach { entry ->
                val emitted = emitter.snippet(VERSION_5, "vm-1", entry.assignment())

                withClue("${entry.version}/${entry.animationId}") {
                    entry.keyframesCss shouldBe "@keyframes ${keyframesName(entry.animationId, entry.version)} { ${entry.body()} }"
                    emitted.lines().first { it.startsWith("@keyframes") } shouldBe entry.keyframesCss
                }
            }
        }

        test("custom properties match in value and in order") {
            fixture.entries.forEach { entry ->
                entry.cases.forEach { case ->
                    val emitted = emitter.snippet(VERSION_5, "vm-1", entry.assignment(case.params))
                    val properties =
                        emitted
                            .lines()
                            .map { it.trim().removeSuffix(";") }
                            .filter { it.startsWith("--") }
                            .associate { line -> line.substringBefore(":") to line.substringAfter(": ") }

                    withClue("${entry.version}/${entry.animationId} (${case.variant})") {
                        properties.keys.toList() shouldContainExactly case.customProperties.keys.toList()
                        properties shouldBe case.customProperties
                    }
                }
            }
        }

        test("Kotlin's split of the verbatim base styles equals TypeScript's parsed map, with no duplicate property") {
            fixture.entries.forEach { entry ->
                val declarations = entry.rawBaseStyles()

                withClue("${entry.version}/${entry.animationId}") {
                    parseDeclarations(declarations) shouldBe entry.baseStyles
                    // A repeated property would make the map smaller than the declaration list.
                    declarations.split(";").count { it.contains(":") } shouldBe entry.baseStyles.size
                }
            }
        }

        test("the shorthand, expanded back to longhands, equals the preview's with the absent ones at their initial values") {
            fixture.entries.forEach { entry ->
                entry.cases.forEach { case ->
                    val emitted = emitter.snippet(VERSION_5, "vm-1", entry.assignment(case.params))
                    val expanded = expandShorthand(shorthandIn(emitted))
                    val expected = INITIAL_LONGHANDS.toMap() + case.longhands

                    withClue("${entry.version}/${entry.animationId} (${case.variant})") {
                        expanded shouldBe expected
                    }
                }
            }
        }

        test("the fixture is not vacuous: every entry carries three cases and the 'all' case moves something") {
            fixture.entries.forEach { entry ->
                withClue("${entry.version}/${entry.animationId}") {
                    entry.cases.map { it.variant } shouldContainExactly listOf("defaults", "all", "partial")
                    val (defaults, all) = entry.cases
                    (all.longhands != defaults.longhands || all.customProperties != defaults.customProperties) shouldBe true
                }
            }
        }
    })

/** The assignment the fixture case describes, on whatever trigger the entry supports. */
private fun ParityEntry.assignment(params: Map<String, String> = emptyMap()) =
    dev.vibemotion.api.domain.Assignment(
        animationId = animationId,
        catalogVersion = version,
        trigger = Trigger.LOAD,
        params = params,
    )

private fun ParityEntry.catalogEntry() =
    checkNotNull(CATALOG.catalog(version)?.entries?.firstOrNull { it.id == animationId }) {
        "The fixture lists $version/$animationId but the bundled catalog does not have it"
    }

private fun ParityEntry.body() = catalogEntry().keyframes

private fun ParityEntry.rawBaseStyles() = catalogEntry().baseStyles.orEmpty()

/**
 * `name duration timing-function delay iteration-count direction fill-mode` back into a longhand
 * map. Split on top-level spaces, because `cubic-bezier(0.16, 1, 0.3, 1)` contains several.
 */
private fun expandShorthand(shorthand: String): Map<String, String> {
    val parts = splitOnTopLevelSpaces(shorthand)
    check(parts.size == INITIAL_LONGHANDS.size + 1) { "Expected 7 shorthand components, got ${parts.size} in '$shorthand'" }
    return buildMap {
        put("animation-name", parts[0])
        INITIAL_LONGHANDS.forEachIndexed { index, (property, _) -> put(property, parts[index + 1]) }
    }
}

private fun splitOnTopLevelSpaces(value: String): List<String> {
    val parts = mutableListOf<String>()
    var depth = 0
    val current = StringBuilder()
    value.forEach { character ->
        when {
            character == '(' -> {
                depth++
                current.append(character)
            }

            character == ')' -> {
                depth--
                current.append(character)
            }

            character == ' ' && depth == 0 -> {
                if (current.isNotEmpty()) parts += current.toString()
                current.clear()
            }

            else -> {
                current.append(character)
            }
        }
    }
    if (current.isNotEmpty()) parts += current.toString()
    return parts
}

File: apps/web/lib/runtime-css/export-parity.test.ts
import { readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { CATALOGS, CATALOG_VERSIONS } from "animation-catalog";
import type { CatalogEntry, CatalogParam } from "animation-catalog";
import { describe, expect, it } from "vitest";

import { assignmentStyle, baseStyleDeclarations, keyframesCss } from "@/lib/runtime-css";

/**
 * The cross-language parity fixture.
 *
 * The preview generator lives here in TypeScript; the exporter is Kotlin. They have to agree on
 * every byte of CSS for every published `(catalogVersion, animationId)` pair, or a designer's
 * export will not be what the designer saw. Nothing but a shared artefact can prove that, so this
 * test *is* the generator: it produces `export-parity.json` from the real `runtime-css` functions,
 * and `apps/api`'s `ExportParityTest` reads the committed file off its test classpath and holds
 * the Kotlin emitter to it.
 *
 * `apps/web` has no TypeScript runner other than vitest, so staleness is a test rather than a
 * script: regenerate deliberately with
 *
 *     VM_UPDATE_PARITY=1 pnpm --filter web test export-parity
 *
 * and read the diff. A change here is a change to what every future export renders.
 */

const here = path.dirname(fileURLToPath(import.meta.url));
const FIXTURE_PATH = path.join(here, "export-parity.json");

type Case = {
  variant: "defaults" | "all" | "partial";
  params: Record<string, string>;
  /** Every `animation-*` longhand `assignmentStyle` writes, `animation-name` included. */
  longhands: Record<string, string>;
  /** Every `--vm-*` custom property, in catalog param order. */
  customProperties: Record<string, string>;
};

type Entry = {
  version: string;
  animationId: string;
  /** The full `@keyframes <name> { … }` block, which the Kotlin emitter must match byte for byte. */
  keyframesCss: string;
  /** TypeScript's *parsed*, ordered declaration map for the entry's `baseStyles`. */
  baseStyles: Record<string, string>;
  cases: Case[];
};

type Fixture = {
  generator: string;
  entries: Entry[];
};

/**
 * A valid value for `param` that is not its default.
 *
 * It has to pass the API's own `paramValueProblem` — the exporter re-validates everything it
 * emits — so each value is taken from what the catalog itself declares: an option, a bound, or a
 * keyword of the right type.
 */
function nonDefaultValue(param: CatalogParam): string {
  switch (param.type) {
    case "select": {
      const options = param.options ?? [];
      return options.find((option) => option !== param.default) ?? param.default;
    }
    case "direction":
      return param.default === "reverse" ? "alternate" : "reverse";
    case "easing":
      return param.default === "linear" ? "ease-in-out" : "linear";
    case "iteration":
      return param.default === "3" ? "4" : "3";
    case "color":
      return param.default === "#112233" ? "#445566" : "#112233";
    default: {
      // The bounded numeric types (duration, length, number, angle, percentage): a bound is by
      // definition inside the range, and the catalog declares both for every one of them.
      if (param.min !== undefined && param.min !== param.default) return param.min;
      if (param.max !== undefined && param.max !== param.default) return param.max;
      return param.default;
    }
  }
}

function casesFor(entry: CatalogEntry): Array<{ variant: Case["variant"]; params: Record<string, string> }> {
  const all: Record<string, string> = {};
  const partial: Record<string, string> = {};
  entry.params.forEach((param, index) => {
    all[param.key] = nonDefaultValue(param);
    // Every other key: a state where some params are stored and the rest fall back to the
    // catalog default, which is what a real assignment usually looks like.
    if (index % 2 === 0) partial[param.key] = nonDefaultValue(param);
  });

  return [
    { variant: "defaults", params: {} },
    { variant: "all", params: all },
    { variant: "partial", params: partial },
  ];
}

function buildFixture(): Fixture {
  const entries: Entry[] = [];

  for (const version of CATALOG_VERSIONS) {
    for (const entry of CATALOGS[version].entries) {
      entries.push({
        version,
        animationId: entry.id,
        keyframesCss: keyframesCss(entry, version),
        baseStyles: baseStyleDeclarations(entry),
        cases: casesFor(entry).map(({ variant, params }) => {
          const style = assignmentStyle(entry, version, params);
          const longhands: Record<string, string> = {};
          const customProperties: Record<string, string> = {};
          for (const [property, value] of Object.entries(style)) {
            if (property.startsWith("animation")) longhands[property] = value;
            else if (property.startsWith("--")) customProperties[property] = value;
          }
          return { variant, params, longhands, customProperties };
        }),
      });
    }
  }

  return {
    generator: "apps/web/lib/runtime-css/export-parity.test.ts — VM_UPDATE_PARITY=1 to regenerate",
    entries,
  };
}

describe("export parity fixture", () => {
  const generated = buildFixture();
  const serialised = `${JSON.stringify(generated, null, 2)}\n`;

  if (process.env.VM_UPDATE_PARITY === "1") {
    writeFileSync(FIXTURE_PATH, serialised);
  }

  it("is committed and up to date with the real runtime-css functions", () => {
    // A stale fixture means the Kotlin exporter is being held to yesterday's TypeScript.
    expect(readFileSync(FIXTURE_PATH, "utf8")).toBe(serialised);
  });

  it("covers every published (version, animationId) pair", () => {
    const expected = CATALOG_VERSIONS.flatMap((version) =>
      CATALOGS[version].entries.map((entry) => `${version}/${entry.id}`),
    );

    expect(generated.entries.map((entry) => `${entry.version}/${entry.animationId}`)).toEqual(expected);
    expect(expected.length).toBeGreaterThan(0);
  });

  it("gives every entry three cases: no params, every param changed, and a partial set", () => {
    for (const entry of generated.entries) {
      expect(entry.cases.map((one) => one.variant)).toEqual(["defaults", "all", "partial"]);

      const [defaults, all] = entry.cases;
      expect(defaults.params).toEqual({});
      const source = CATALOGS[entry.version as keyof typeof CATALOGS].entries.find((e) => e.id === entry.animationId);
      expect(Object.keys(all.params)).toEqual(source?.params.map((param) => param.key));
    }
  });

  it("changes something for every param in the 'all' case", () => {
    for (const entry of generated.entries) {
      const source = CATALOGS[entry.version as keyof typeof CATALOGS].entries.find((e) => e.id === entry.animationId);
      const [defaults, all] = entry.cases;

      for (const param of source?.params ?? []) {
        const property = param.cssVar ?? undefined;
        const before = property ? defaults.customProperties[property] : undefined;
        const after = property ? all.customProperties[property] : undefined;
        if (property) expect(after).not.toBe(before);
      }
      // Standard params move too, or the longhand halves of the comparison would be vacuous.
      expect(all.longhands).not.toEqual(defaults.longhands);
    }
  });

  it("writes every animation longhand the preview sets, animation-name included", () => {
    for (const entry of generated.entries) {
      for (const one of entry.cases) {
        expect(one.longhands["animation-name"]).toBeTruthy();
        for (const property of Object.keys(one.longhands)) {
          expect(property.startsWith("animation")).toBe(true);
        }
        for (const property of Object.keys(one.customProperties)) {
          expect(property.startsWith("--vm-")).toBe(true);
        }
      }
    }
  });
});
